src/lib.rs

```diff
@@ -1774,66 +1774,39 @@ impl Build {
+        // Delete the destination if it exists as we want to
+        // create on the first iteration instead of appending.
         let _ = fs::remove_file(&dst);
-
-        let objects: Vec<_> = objs.iter().map(|obj| obj.dst.clone()).collect();
         let target = self.get_target()?;
-        if target.contains("msvc") {
-            let (mut cmd, program) = self.get_ar()?;
-            let mut out = OsString::from("-out:");
-            out.push(dst);
-            cmd.arg(out).arg("-nologo");
-            for flag in self.ar_flags.iter() {
-                cmd.arg(flag);
-            }
-
-            // Similar to https://github.com/rust-lang/rust/pull/47507
-            // and https://github.com/rust-lang/rust/pull/48548
-            let estimated_command_line_len = objects
-                .iter()
-                .chain(&self.objects)
-                .map(|a| a.as_os_str().len())
-                .sum::<usize>();
-            if estimated_command_line_len > 1024 * 6 {
-                let mut args = String::from("\u{FEFF}"); // BOM
-                for arg in objects.iter().chain(&self.objects) {
-                    args.push('"');
-                    for c in arg.to_str().unwrap().chars() {
-                        if c == '"' {
-                            args.push('\\')
-                        }
-                        args.push(c)
-                    }
-                    args.push('"');
-                    args.push('\n');
-                }
-
-                let mut utf16le = Vec::new();
-                for code_unit in args.encode_utf16() {
-                    utf16le.push(code_unit as u8);
-                    utf16le.push((code_unit >> 8) as u8);
-                }
-
-                let mut args_file = OsString::from(dst);
-                args_file.push(".args");
-                fs::File::create(&args_file)
-                    .unwrap()
-                    .write_all(&utf16le)
-                    .unwrap();
-
-                let mut args_file_arg = OsString::from("@");
-                args_file_arg.push(args_file);
-                cmd.arg(args_file_arg);
+        let objs: Vec<_> = objs
+            .iter()
+            .map(|o| o.dst.clone())
+            .chain(self.objects.clone())
+            .collect();
+
+        let chunk_size = if self.get_host()?.contains("msvc") {
+            let max_command_line_len = 1024 * 6;
+            let estimated_command_line_len =
+                objs.iter().map(|a| a.as_os_str().len()).sum::<usize>();
+            if estimated_command_line_len > max_command_line_len {
+                let average_chars = estimated_command_line_len / objs.len();
+                max_command_line_len / average_chars
             } else {
-                cmd.args(&objects).args(&self.objects);
+                objs.len()
             }
-            run(&mut cmd, &program)?;
+        } else {
+            objs.len()
+        };
 
+        for chunk in objs.chunks(chunk_size) {
```

Comment:
Instead of calculating `chunk_size` could we perhaps conservatively just say the chunk size is 100? That seems like it would probably cover almost all use cases I think?

Reply:
Sure thing!